api/src/main/java/org/infinispan/api/async/Cache.java

```diff
@@ -0,0 +1,134 @@
+package org.infinispan.api.async;
+
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.infinispan.api.common.AsyncIterator;
+import org.infinispan.api.common.KeyValueEntry;
+import org.infinispan.api.common.events.KeyValueListener;
+import org.infinispan.api.common.events.ListenerHandle;
+import org.infinispan.api.configuration.CacheConfiguration;
+
+/**
+ * @since 13.0
+ **/
+public interface Cache<K, V> {
+   /**
+    * Get the value of the Key if such exists
+    *
+    * @param key
+    * @return the value
+    */
+   CompletionStage<V> get(K key);
+
+   /**
+    * Insert the key/value if such key does not exist
+    *
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<Boolean> putIfAbsent(K key, V value);
+
+   /**
+    * Save the key/value. If the key exists will replace the value and the previous value will be returned.
+    *
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<V> put(K key, V value);
+
+
+   CompletionStage<Void> set(K key, V value);
+
+   /**
+    * Delete the key
+    *
+    * @param key
+    * @return whether the entry was removed.
+    */
+   CompletionStage<Boolean> remove(K key);
+
+   /**
+    * Retrieve all keys
+    *
+    * @return
+    */
+   AsyncIterator<K> keys();
+
+   /**
+    * Retrieve all entries
+    *
+    * @return
+    */
+   AsyncIterator<? extends Map.Entry<K, V>> entries();
+
+   /**
+    * @param entries
+    * @return Void
+    */
+   CompletionStage<Void> putAll(Map<K, V> entries);
+
+   /**
+    * Estimate the size of the store
+    *
+    * @return Long, estimated size
+    */
+   CompletionStage<Long> estimateSize();
+
+   /**
+    * Clear the store. If a concurrent operation puts data in the store the clear might not properly work
+    *
+    * @return Void
+    */
+   CompletionStage<Void> clear();
+
+   /**
+    * Executes the query and returns an iterable with the entries
+    *
+    * @param query query String
+    * @return Publisher reactive streams
+    */
+   CompletionStage<Iterable<KeyValueEntry<K, V>>> find(String query);
```

Comment:
Query results can potentially contain other metadata such as  ```time``` and each entry potentially can return a full-text score (we don't have it now, but may be added in the future) and other associated info, so it's better to contemplate it. currently we have  ````QueryResult<T> execute();```

Reply:
Yes, that was a quick translation from the previous `KeyValueStore.find()`.

Reply:
Done, but I'm sure this will change further after friday's chat